### PR TITLE
Add linear constraints to LLSModel

### LINKED ISSUE #100: Implement constraints in NLSModels
TODO:
- [x] ADNLSModel (#93)
- [x] SimpleNLSModel (#102)
- [ ] LLSModel (#103)
- [x] Add consistency checks (#102)

### PR DESCRIPTION
**Pull request by abelsiqueira**
Closes #100

### CODE REVIEW

File: src/lls_model.jl [hunk 1]
```diff
@@ -3,9 +3,10 @@ export LLSModel,
        jac_op_residual, hess_residual, hprod_residual!
 
 """
-    nls = LLSModel(A, b)
+    nls = LLSModel(A, b; lvar, uvar, C, lcon, ucon)
 
-Creates a Linear Least Squares model ½‖Ax - b‖²
+Creates a Linear Least Squares model ½‖Ax - b‖² with optional bounds
+`lvar ≦ x ≦ y` and optional linear constraints `lcon ≦ C ≦ ucon`.
```
Comment (dpo) [inline src/lls_model.jl]: `lcon ≤ Cx ≤ ucon` ?

File: src/lls_model.jl [hunk 2]
```diff
@@ -14,21 +15,32 @@ type LLSModel <: AbstractNLSModel
...
+                  ucon :: Vector = Float64[],
+                  y0 :: Vector = zeros(size(C,1)))
   m, n = size(A)
   if length(b) != m
-    error("Incompatibility detected: A is $m×$n and b has lenght $(length(b))")
+    error("Incompatibility detected: A is $m×$n and b has length $(length(b))")
   end
+  ncon = size(C, 1)
+  if !(ncon == length(lcon) == length(ucon) == length(y0))
+    error("The number of lines in C must be the same length as lcon, ucon and y0")
```
Comment (dpo) [inline src/lls_model.jl]: lines <- rows

File: src/lls_model.jl [hunk 3]
```diff
@@ -69,3 +81,83 @@ function hprod_residual!(nls :: LLSModel, x :: AbstractVector, i :: Int, v :: Ab
...
+  c[1:nls.meta.ncon] = nls.C * x
+  return c
+end
+
+function jac_coord(nls :: LLSModel, x :: Vector)
+  increment!(nls, :neval_jac)
+  if isa(nls.C, AbstractMatrix)
+    return findnz(J)
+  else
+    return findnz(full(J))
```
Comment (dpo) [inline src/lls_model.jl]: Is that really what we want? Linear operators should probably not be used that way. If constraints are defined by an operator, it means we shouldn't perform matrix operations.

Comment (abelsiqueira) [reply src/lls_model.jl]: You're right, that makes more sense.

File: src/lls_model.jl [hunk 4]
```diff
@@ -95,11 +95,10 @@ end
 
 function jac_coord(nls :: LLSModel, x :: Vector)
   increment!(nls, :neval_jac)
-  if isa(nls.C, AbstractMatrix)
-    return findnz(J)
-  else
-    return findnz(full(J))
+  if isa(nls.C, LinearOperator)
+    error("hess is not defined for LinearOperators")
```
Comment (dpo) [inline src/lls_model.jl]: Hess or jac?

Comment (abelsiqueira) [reply src/lls_model.jl]: Fixed

**Comment by dpo** [comment]
Thank you!

**Pull request merged**